src/libs/xapi-stats.ts

```diff
@@ -0,0 +1,438 @@
+        )
+    }
+    const currentTimeStamp = await this.#xapi.getHostServertime(host)
+    
+    const stats = this.#getCachedStats(uuid, step, currentTimeStamp)
+    if (stats !== undefined) {
+        return stats
+    }
+    
+    const maxDuration = step * RRD_POINTS_PER_STEP[step]
+    
+    // To avoid crossing over the boundary, we ask for one less step
+    const optimumTimestamp = currentTimeStamp - maxDuration + step
+    const json = await this._getJson(host, optimumTimestamp, step)
+    
+    if(uuid === "1d381a66-d1cb-bb7e-50a1-feeab58b293d") console.log("VM")
+    console.log(json)
+    const actualStep = json.meta.step
+  
+    if (json.data.length > 0) {
+      // fetched data is organized from the newest to the oldest
+      // but this implementation requires it in the other direction
+      json.data.reverse()
+      json.meta.legend.forEach((legend: any, index: number) => {
+        const [, type, uuid, metricType] = /^AVERAGE:([^:]+):(.+):(.+)$/.exec(legend) as any
+
+        const metrics = STATS[type] as any
+        if (metrics === undefined) {
+          return
+        }
+
+        const { metric, testResult } = findMetric(metrics, metricType) as any
+        if (metric === undefined) {
+          return
+        }
+
+        const xoObjectStats = createGetProperty(this.#statsByObject, uuid, {})
+        let stepStats = xoObjectStats[actualStep]
+        if (stepStats === undefined || stepStats.endTimestamp !== json.meta.end) {
+          stepStats = xoObjectStats[actualStep] = {
+            endTimestamp: json.meta.end,
+            interval: actualStep,
+          }
+        }
+
+        const path = metric.getPath !== undefined ? metric.getPath(testResult) : [findKey(metrics, metric)]
+
+        const lastKey = path.length - 1
+        let metricStats = createGetProperty(stepStats, 'stats', {})
+        path.forEach((property: any, key: number) => {
+          if (key === lastKey) {
+            metricStats[property] = computeValues(json.data, index, metric.transformValue)
+            return
+          }
+
+          metricStats = createGetProperty(metricStats, property, {})
+        })
+      })
+    }
+
+    if (actualStep !== step) {
+      throw new FaultyGranularity(`Unable to get the true granularity: ${actualStep}`)
+    }
+
+    return (
+      this.#statsByObject[uuid]?.[step] ?? {
+        endTimestamp: currentTimeStamp,
+        interval: step,
+        stats: {},
+      }
+    )
+  }
+
+  getHostStats = (hostId: string, granularity: Granularity): Promise<XapiStatsResponse<HostStats>> => {
+    const host = useHostStore().getRecordByUuid(hostId)
+    if (host === undefined) {
+      throw new Error(`Host ${hostId} could not be found.`)
+    }
+    return this._getAndUpdateStats({
+        host,
+        uuid: host.uuid,
+        granularity
+    })
+  }
+
+  getVmStats = (vmId: string, granularity: Granularity): Promise<XapiStatsResponse<VmStats>> => {
```

Comment:
since it's VM specific , do you think it should go to the vm store ?

Reply:
Despite the name of "store", I think actions related to a specific type should be defined in the according store.